### PR TITLE
feat: Add variable to allow userdata to be passed as string to aws launch template

### PR DESCRIPTION
**Pull request by rashidnhm**
Fix for the issue mentioned in #1960. This PR enables the passing of userdata script as a string to the module.

If both variables `userdata_string` and `userdata_template` are passed, then `userdata_string` would take precedence.

### CODE REVIEW

**Comment by npalm** [comment]
Note: check conflicts with #1956

**Comment by ScottGuymer** [comment]
As this is a quite advanced usecase I think this needs an example to show a potential user how to do this and call out some of the things they need to consider along the way.

File: modules/runners/variables.tf [hunk 1]
```diff
@@ -144,6 +144,12 @@ variable "userdata_template" {
   default     = null
 }
 
+variable "userdata_string" {
+  description = "Alternative user-data as string, replacing the default template. By providing your own user_data you have to take care of installing all required software, including the action runner. Variables userdata_pre/post_install and userdata_template are ignored."
```
Comment (ScottGuymer) [inline modules/runners/variables.tf:148]: It needs to be made clear in this that this is an advanced use case and here be dragons. Maybe prepend `**Advanced**` or something, not sure if we have a convention for this yet.

It also needs an example with an extensive readme on what to look for and the fact that future upgrades can likely cause issues.

**Review by ScottGuymer — CHANGES_REQUESTED** [review]

File: main.tf [hunk 2]
```diff
@@ -188,6 +188,7 @@ module "runners" {
 
   enabled_userdata      = var.enabled_userdata
   userdata_template     = var.userdata_template
+  userdata_string       = var.userdata_string
```
[the remark this answers is not in the document]
Comment (rashidnhm) [reply main.tf]: I have updated the variable name to `userdata` as I felt it captured the usecase well. I have also updated the variable description to include more "use with care" cautions.

File: modules/runners/variables.tf (hunk 1 above)
Comment (rashidnhm) [reply modules/runners/variables.tf]: Variable description has been updated. Please re-assess and let me know

**Comment by rashidnhm** [comment]
I am also open to adding a `userdata` example if you'd like. Let me know.

**Comment by ScottGuymer** [comment]
Yes, I think an example of how to use this is warranted. With some notes in the readme to guide the user when they might want to use it and what issues there could be.

File: main.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (npalm) [reply main.tf]: I think userdata_override is a better name; it gives a much clearer message it is ovrriding the default behavior of the module. By overriding the user_data the user of the module have to write a own custom script to manage the runner instllation and registration.

**Review by npalm — CHANGES_REQUESTED** [review]
@rashidnhm please can you rebase the PR. We have updated the module last week to AWS terraform provide 4.x

File: main.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (rashidnhm) [reply main.tf]: Sounds good, I will update shortly

[the remark this answers is not in the document]
Comment (rashidnhm) [reply main.tf]: I have updated the userdata variable name to `userdata_override`

File: README.md [hunk 3]
```diff
@@ -477,6 +477,7 @@ In case the setup does not work as intended follow the trace of events:
...
 | <a name="input_userdata_pre_install"></a> [userdata\_pre\_install](#input\_userdata\_pre\_install) | Script to be ran before the GitHub Actions runner is installed on the EC2 instances | `string` | `""` | no |
 | <a name="input_userdata_template"></a> [userdata\_template](#input\_userdata\_template) | Alternative user-data template, replacing the default template. By providing your own user\_data you have to take care of installing all required software, including the action runner. Variables userdata\_pre/post\_install are ignored. | `string` | `null` | no |
+| <a name="input_userdata"></a> [userdata](#input\_userdata) | Alternative to userdata\_template to pass userdata as string. Note: Usage of this variable overrides all templates in place by this module. You are responsible for the setup and installation of all software, including the action runner. Use with caution. Variables userdata\_pre/post\_install and userdata\_template are ignored. | `string` | `null` | no |
```
Comment (ScottGuymer) [inline README.md:480]: This doc just needs updating to the new name.

File: modules/runners/variables.tf (hunk 1 above)
Comment (ScottGuymer) [reply modules/runners/variables.tf]: LGTM

**Comment by ScottGuymer** [comment]
I think we are just missing the example now.

File: README.md (hunk 3 above)
Comment (rashidnhm) [reply README.md]: I've updated the README

**Comment by npalm** [comment]
Will put it on my list, sorry the PR got out of sight.

**Comment by rashidnhm** [comment]
@npalm Sorry about the late reply, I've fixed the formatting issue. Please take a look and let me know.

**Comment by rashidnhm** [comment]
@npalm Just checking in to see if you got a chance to comb through this?

**Comment by npalm** [comment]
not yet, hope to check all the open PR's this week

**Comment by rashidnhm** [comment]
@npalm I will try to update this PR soon and pull in all latest changes, will you have time to review anytime soon?

**Comment by rashidnhm** [comment]
@npalm the reason I had this PR is to add functionality myself for having multiple runners. However, now that #2472 is in progress, this is no longer needed on my end.

**Comment by npalm** [comment]
I think in that case we should close, thanks fro reaching out

**Pull request closed without merging**